src/core/DY.Core.Label.fst

```diff
@@ -318,3 +292,89 @@ let join_flow_to_public_eq tr x1 x2 =
   reveal_opaque (`%can_flow) (can_flow);
   reveal_opaque (`%join) (join);
   reveal_opaque (`%public) (public)
+
+/// Generic label for states:
+/// `state_pred_label p` is corrupt when
+/// there exists a corrupt `SetState` event satisfying the predicate `p`.
+/// It can for example be used to depict any state of a principal
+/// (e.g. as done by `principal_label`)
+
+[@@"opaque_to_smt"]
+val state_pred_label:
+  (principal -> state_id -> bytes -> prop) ->
+  label
+let state_pred_label p = mk_label {
+  is_corrupt = (fun tr ->
+    exists time prin sess_id content.
+      event_exists tr (Corrupt time) /\
+      event_at tr time (SetState prin sess_id content) /\
+      p prin sess_id content
+  );
+  is_corrupt_later = (fun tr1 tr2 -> ());
+}
+
+val state_pred_label_can_flow:
+  (principal -> state_id -> bytes -> prop) ->
+  (principal -> state_id -> bytes -> prop) ->
+  prop
+let state_pred_label_can_flow p1 p2 =
+  forall p s c. p2 p s c ==> p1 p s c
+
+val state_pred_label_can_flow_state_pred_label:
+  tr:trace ->
+  p1:(principal -> state_id -> bytes -> prop) -> p2:(principal -> state_id -> bytes -> prop) ->
+  Lemma
+  (requires state_pred_label_can_flow p1 p2)
+  (ensures state_pred_label p1 `can_flow tr` state_pred_label p2)
+  [SMTPat (state_pred_label p1 `can_flow tr` state_pred_label p2)]
+let state_pred_label_can_flow_state_pred_label tr p1 p2 =
+  reveal_opaque (`%is_corrupt) is_corrupt;
+  reveal_opaque (`%can_flow) (can_flow);
+  reveal_opaque (`%state_pred_label) (state_pred_label)
+
+val state_pred_label_can_flow_public:
+  tr:trace ->
+  p:(principal -> state_id -> bytes -> prop) ->
+  time:timestamp -> prin:principal -> sess_id:state_id -> content:bytes ->
+  Lemma
+  (requires (
+    event_exists tr (Corrupt time) /\
+    event_at tr time (SetState prin sess_id content) /\
+    p prin sess_id content
+  ))
+  (ensures (state_pred_label p) `can_flow tr` public)
```

Comment:
This should probably be changed to a `<==>` lemma type, as this one doesn't help saying what happened when `state_pred_label p` flows to public (which can be useful in e.g. security theorems?)

Reply:
Done